### PR TITLE
feat(lambda): store lambda scripts within the dgraph, send it to JS s…

### PR DESCRIPTION
**Pull request by NamanJain8**
Related to https://github.com/dgraph-io/dgraph-lambda/pull/22

This PR changes the way lambda is handled by Dgraph. Earlier, the user had to start the server with a predefined lambda script. This does not scale well in a multi-tenant environment. As we will need N servers (or endpoints) to handle the load. Also, each time the script is changed, the lambda server needs to be restarted.

With this change, instead of loading the script from the predefined file, we will be sending the script in the request body to the lambda-server.

The details are mentioned in the [discuss post](https://discuss.dgraph.io/t/rfc-storing-lambda-script-in-dgraph/14972). To summarize:
### Storage
We will store the lambda script within `dgraph.graphql.schema` predicate. The value will be JSON marshal of
```
type GQL struct {
   Script:     String
   Schema: String
}
```
This change has been done in a backward-compatible manner. 
### New resolvers
This adds 2 new admin resolvers:
- UpdateLambdaScript → accepts base64 encoded lambda script
- GetLambdaScript → Returns base64 encoded script

### Export/Import:
The GraphQL schema is exported in a separate file in the following JSON format:
[ExportedGQL] where ExportedGQL type is {Namespace: Int, Schema: String}. We will extend it to {Namespace: Int, Schema: String, Script: String} and hence the bulk loader can handle it properly.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7955)

### CODE REVIEW

**Review by manishrjain — COMMENTED** [review]
:lgtm: Had a rough look, looks good. Get a deeper review by @aman-bansal / @ahsanbarkati 

> Reviewed 12 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-:-MezcK599FQF7RQQw21s:byhhs4i)* status: all files reviewed, 1 unresolved discussion (waiting on @NamanJain8)

___
*[protos/pb.proto, line 743 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-MezcDY70ylJi_bZfVOA:-MezcDY70ylJi_bZfVOB:bqyc69i) ([raw file](https://github.com/dgraph-io/dgraph/blob/f992c27d3317fa09f152364366f42f5521fd6608/protos/pb.proto#L743)):*
> ```protobuf
> message UpdateGraphQLSchemaRequest {
>   enum Op {
>     SCHEMA = 0;
> ```

Do you need the Op?

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7955)* status: all files reviewed, 1 unresolved discussion (waiting on @manishrjain)

___
*[protos/pb.proto, line 743 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-MezcDY70ylJi_bZfVOA:-Mezcf_HEqo0xFs-rP14:b-vr57y4) ([raw file](https://github.com/dgraph-io/dgraph/blob/f992c27d3317fa09f152364366f42f5521fd6608/protos/pb.proto#L743)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Do you need the Op?
</blockquote></details>

Op is needed because update the schema to an empty string, then we won't be able to decide the type of update on the receiver.

File: graphql/e2e/directives/docker-compose.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ services:
       --trace "ratio=1.0;"
 
   lambda:
-    image: dgraph/dgraph-lambda:latest
+    image: namanj1811/lambda-test:v2
```
Comment (NamanJain8) [inline graphql/e2e/directives/docker-compose.yml:40]: TODO: I will update this once https://github.com/dgraph-io/dgraph-lambda/pull/22 is merged.

File: graphql/e2e/normal/docker-compose.yml [hunk 2]
```diff
@@ -37,7 +37,7 @@ services:
       --trace "ratio=1.0;"
 
   lambda:
-    image: dgraph/dgraph-lambda:latest
+    image: namanj1811/lambda-test:v1
```
Comment (NamanJain8) [inline graphql/e2e/normal/docker-compose.yml:40]: TODO: I will update this once https://github.com/dgraph-io/dgraph-lambda/pull/22 is merged.

**Review by ahsanbarkati — APPROVED** [review]
:lgtm: Looks good to me. Few nits. 

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-:-MfbxCa45QSeWQ55UcfX:b-k5fhad)* status: 8 of 20 files reviewed, 5 unresolved discussions (waiting on @manishrjain and @NamanJain8)

___
*[edgraph/server.go, line 500 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-MfbqnXSE4wPo8WeZ7SW:-MfbqnXSE4wPo8WeZ7SX:bt7lqee) ([raw file](https://github.com/dgraph-io/dgraph/blob/c78ae78b2550d2d9bf7c3b1f8c09dde49d894353/edgraph/server.go#L500)):*
> ```Go
> 		// just reinsert the GraphQL schema, no need to alter dgraph schema as this was drop_data
> 		if _, e := UpdateGQLSchema(ctx, graphQLSchema, ""); e != nil {
> 			err = errors.Errorf("While updating gql schema, got err: %+v.", e)
> ```

If `e` is not `nil` then why not return here itself? Why do we need to accumulate the errors?
___
*[worker/lambda_script.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-MfbwZ3z-w8VVWOHhEoF:-MfbwZ3z-w8VVWOHhEoG:b-us4g3y) ([raw file](https://github.com/dgraph-io/dgraph/blob/c78ae78b2550d2d9bf7c3b1f8c09dde49d894353/worker/lambda_script.go#L31)):*
> ```Go
> 
> type LambdaScriptStore struct {
> 	mux    sync.RWMutex
> ```

Can we avoid the variable mux, it will make it look cleaner. 
```
type LambdaScriptStore struct {
  sync.RQMutex
  ...
}
```
Then directly do `ls.Lock()` instead of `ls.mux.Lock()`

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7955)* status: 7 of 20 files reviewed, 5 unresolved discussions (waiting on @ahsanbarkati and @manishrjain)

___
*[edgraph/server.go, line 500 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-MfbqnXSE4wPo8WeZ7SW:-Mfc0aLb4bpbY0LFabz0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c78ae78b2550d2d9bf7c3b1f8c09dde49d894353/edgraph/server.go#L500)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

If `e` is not `nil` then why not return here itself? Why do we need to accumulate the errors?
</blockquote></details>

Done.
___
*[worker/lambda_script.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7955#-MfbwZ3z-w8VVWOHhEoF:-Mfc0edR-CSU4VPFeu3O:boyexbg) ([raw file](https://github.com/dgraph-io/dgraph/blob/c78ae78b2550d2d9bf7c3b1f8c09dde49d894353/worker/lambda_script.go#L31)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Can we avoid the variable mux, it will make it look cleaner. 
```
type LambdaScriptStore struct {
  sync.RQMutex
  ...
}
```
Then directly do `ls.Lock()` instead of `ls.mux.Lock()`
</blockquote></details>

Done. Thanks.

**Pull request merged**